@grpc/grpc-js is incompatible with typescript definations @types/node@17.0.6

### Problem description
when project already has the dependency `@types/node@17.0.6`, `tsc` failed to build

### Reproduction steps

- `yarn add @types/node@17.0.6 && tsc -b`
- failed with output: 
```
node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18` - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
  The types returned by 'end(...)' are incompatible between these types.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: () => void): this; (chunk: any, cb?: () => void): this; (chunk: any, encoding: BufferEncoding, cb?: () => void): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

64     end(metadata?: any): void;
       ~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: () => void): this; (chunk: any, cb?: () => void): this; (chunk: any, encoding?: BufferEncoding, cb?: () => void): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

77     end(metadata?: any): void;
       ~~~


Found 3 errors.
```
- `yarn add @types/node@17.0.5 && tsc -b`
- successfully built again.

### Environment
 - OS name, version and architecture: M1 Pro, Macbook Pro
 - Node version: 16.13.0
 - Node installation method: brew
 - Package name and version: `"@grpc/grpc-js": "^1.4.5"`

### tsconfig.json
`{
  "compilerOptions": {
    "lib": [
      "ES2021"
    ],
    "module": "commonjs",
    "moduleResolution": "node",
    "target": "ES2021",
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "outDir": "dist"
  },
  "include": [
    "src/**/*"
  ],
  "exclude": [
    "node_modules",
    "**/*.spec.ts"
  ]
}`

It seems that the version of dependency package `@types/node` should be limited, rather than current ">=12.12.47".

Same issue for me, buy my @types/node is 14.17.0 so it doesn't feel like it's an issue with node types themselves ... Error I'm getting is:

```
remote:        node_modules/@grpc/grpc-js/build/src/object-stream.d.ts(11,18): error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
remote:          The types returned by 'end(...)' are incompatible between these types.
remote:            Type 'void' is not assignable to type 'this'.
remote:              'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
remote:        node_modules/@grpc/grpc-js/build/src/server-call.d.ts(64,5): error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
remote:          Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
remote:            Type 'void' is not assignable to type 'this'.
remote:              'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
remote:        node_modules/@grpc/grpc-js/build/src/server-call.d.ts(77,5): error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
remote:          Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.
remote:            Type 'void' is not assignable to type 'this'.
remote:              'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
```

Still working on a fix ... unable to deploy production app.

Interesting. It does seem to be a problem with `@types/node@14.18.4`. I just created a standalone test project to troubleshoot this. By default, `@types/node@14.18.4` was used. This was my default output for `npm ls @types/node`:

```
├─┬ @google-cloud/compute@3.0.0
│ └─┬ google-gax@2.28.1
│   ├─┬ @grpc/grpc-js@1.4.5
│   │ └── @types/node@14.18.4 deduped
│   └─┬ protobufjs@6.11.2
│     └── @types/node@14.18.4 deduped
├─┬ @types/express@4.17.13
│ ├─┬ @types/body-parser@1.19.2
│ │ ├─┬ @types/connect@3.4.35
│ │ │ └── @types/node@14.18.4 deduped
│ │ └── @types/node@14.18.4 deduped
│ ├─┬ @types/express-serve-static-core@4.17.27
│ │ └── @types/node@14.18.4 deduped
│ └─┬ @types/serve-static@1.13.10
│   └── @types/node@14.18.4 deduped
├── @types/node@14.18.4
└─┬ firebase-admin@9.12.0
  └── @types/node@14.18.4 deduped
```

But when I tried to compile with `tsc`, I got the error described in this issue:

```
> tsc

  The types returned by 'end(...)' are incompatible between these types.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

64     end(metadata?: any): void;
       ~~~

  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

77     end(metadata?: any): void;
       ~~~


Found 3 errors.
```

But now when I force `@types/node@14.18.3` as @darioxtx suggested, everything works. Here's the output of `npm ls @types/node` after the fix:

```
├─┬ @google-cloud/compute@3.0.0
│ └─┬ google-gax@2.28.1
│   ├─┬ @grpc/grpc-js@1.4.5
│   │ └── @types/node@14.18.3 deduped
│   └─┬ protobufjs@6.11.2
│     └── @types/node@14.18.3 deduped
├─┬ @types/express@4.17.13
│ ├─┬ @types/body-parser@1.19.2
│ │ ├─┬ @types/connect@3.4.35
│ │ │ └── @types/node@14.18.3 deduped
│ │ └── @types/node@14.18.3 deduped
│ ├─┬ @types/express-serve-static-core@4.17.27
│ │ └── @types/node@14.18.3 deduped
│ └─┬ @types/serve-static@1.13.10
│   └── @types/node@14.18.3 deduped
├── @types/node@14.18.3
└─┬ firebase-admin@9.12.0
  └── @types/node@14.18.3 deduped
```

I am getting the below error 

node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18 - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
  The types returned by 'end(...)' are incompatible between these types.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

64     end(metadata?: any): void;
       ~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: "ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

77     end(metadata?: any): void;
       ~~~
Found 3 errors.
Error: functions predeploy error: Command terminated with non-zero exit code2


npm ls @types/node

functions@ ....

+-- firebase-admin@10.0.1
| +-- @google-cloud/firestore@4.15.1
| | +-- google-gax@2.28.1
| | | `-- @grpc/grpc-js@1.4.5
| | |   `-- @types/node@17.0.6 deduped
| | `-- protobufjs@6.11.2
| |   `-- @types/node@17.0.6 deduped
| `-- @types/node@17.0.6
`-- firebase-functions@3.16.0
  `-- @types/express@4.17.3
    +-- @types/body-parser@1.19.2
    | +-- @types/connect@3.4.35
    | | `-- @types/node@17.0.6 deduped
    | `-- @types/node@17.0.6 deduped
    +-- @types/express-serve-static-core@4.17.27
    | `-- @types/node@17.0.6 deduped
    `-- @types/serve-static@1.13.10
      `-- @types/node@17.0.6 deduped

Hi @johnnyoshika - I changed the @types/node to 14.18.3 as you suggested, but I'm still getting the error every time I attemtp to `npm run build`
```
node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18 - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
  The types returned by 'end(...)' are incompatible between these types.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

11 export interface IntermediateObjectWritable<T> extends Writable {
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

64     end(metadata?: any): void;
       ~~~

node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
  Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: "ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

77     end(metadata?: any): void;
       ~~~

src/index.ts:1:1 - error TS6133: 'functions' is declared but its value is never read.

1 import * as functions from "firebase-functions";
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 4 errors.

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! functions@ build: `tsc`
npm ERR! Exit status 2
```

> Hi @johnnyoshika - I changed the @types/node to 14.18.3 as you suggested, but I'm still getting the error every time I attemtp to `npm run build` (after that I did the `yarn install`, of course)
> 
> ```
> node_modules/@grpc/grpc-js/build/src/object-stream.d.ts:11:18 - error TS2430: Interface 'IntermediateObjectWritable<T>' incorrectly extends interface 'Writable'.
>   The types returned by 'end(...)' are incompatible between these types.
>     Type 'void' is not assignable to type 'this'.
>       'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
> 
> 11 export interface IntermediateObjectWritable<T> extends Writable {
>                     ~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> node_modules/@grpc/grpc-js/build/src/server-call.d.ts:64:5 - error TS2416: Property 'end' in type 'ServerWritableStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Writable'.
>   Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding: BufferEncoding, cb?: (() => void) | undefined): this; }'.
>     Type 'void' is not assignable to type 'this'.
>       'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
> 
> 64     end(metadata?: any): void;
>        ~~~
> 
> node_modules/@grpc/grpc-js/build/src/server-call.d.ts:77:5 - error TS2416: Property 'end' in type 'ServerDuplexStreamImpl<RequestType, ResponseType>' is not assignable to the same property in base type 'Duplex'.
>   Type '(metadata?: any) => void' is not assignable to type '{ (cb?: (() => void) | undefined): this; (chunk: any, cb?: (() => void) | undefined): this; (chunk: any, encoding?: "ascii" | "utf8" | "utf-8" | "utf16le" | "ucs2" | "ucs-2" | "base64" | "base64url" | "latin1" | "binary" | "hex" | undefined, cb?: (() => void) | undefined): this; }'.
>     Type 'void' is not assignable to type 'this'.
>       'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
> 
> 77     end(metadata?: any): void;
>        ~~~
> 
> src/index.ts:1:1 - error TS6133: 'functions' is declared but its value is never read.
> 
> 1 import * as functions from "firebase-functions";
>   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> 
> Found 4 errors.
> 
> npm ERR! code ELIFECYCLE
> npm ERR! errno 2
> npm ERR! functions@ build: `tsc`
> npm ERR! Exit status 2
> ```

Are you using both yarn and npm? When using yarn 1, I still had the non-working 17.0.6 types in the grpc module even after changing to 14x in the main package. Probably bc. of defective dedupe in yarn1. Try and see yourself with ‘yarn list —pattern @types/node‘.
I solved it by finally upgrading to yarn3.

--- a/packages/grpc-js/package.json
+++ b/packages/grpc-js/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@grpc/grpc-js",
-  "version": "1.4.5",
+  "version": "1.4.6",
   "description": "gRPC Library for Node - pure JS implementation",
   "homepage": "https://grpc.io/",
   "repository": "https://github.com/grpc/grpc-node/tree/master/packages/grpc-js",

--- a/packages/grpc-js/src/object-stream.ts
+++ b/packages/grpc-js/src/object-stream.ts
@@ -36,9 +36,9 @@ export interface IntermediateObjectWritable<T> extends Writable {
   write(chunk: any & T, cb?: WriteCallback): boolean;
   write(chunk: any & T, encoding?: any, cb?: WriteCallback): boolean;
   setDefaultEncoding(encoding: string): this;
-  end(): this;
-  end(chunk: any & T, cb?: Function): this;
-  end(chunk: any & T, encoding?: any, cb?: Function): this;
+  end(): ReturnType<Writable['end']> extends Writable ? this : void;
+  end(chunk: any & T, cb?: Function): ReturnType<Writable['end']> extends Writable ? this : void;
+  end(chunk: any & T, encoding?: any, cb?: Function): ReturnType<Writable['end']> extends Writable ? this : void;
 }
 
 export interface ObjectWritable<T> extends IntermediateObjectWritable<T> {
@@ -46,7 +46,7 @@ export interface ObjectWritable<T> extends IntermediateObjectWritable<T> {
   write(chunk: T, cb?: Function): boolean;
   write(chunk: T, encoding?: any, cb?: Function): boolean;
   setDefaultEncoding(encoding: string): this;
-  end(): this;
-  end(chunk: T, cb?: Function): this;
-  end(chunk: T, encoding?: any, cb?: Function): this;
+  end(): ReturnType<Writable['end']> extends Writable ? this : void;
+  end(chunk: T, cb?: Function): ReturnType<Writable['end']> extends Writable ? this : void;
+  end(chunk: T, encoding?: any, cb?: Function): ReturnType<Writable['end']> extends Writable ? this : void;
 }

--- a/packages/grpc-js/src/object-stream.ts
+++ b/packages/grpc-js/src/object-stream.ts
@@ -36,9 +36,9 @@ export interface IntermediateObjectWritable<T> extends Writable {
   write(chunk: any & T, cb?: WriteCallback): boolean;
   write(chunk: any & T, encoding?: any, cb?: WriteCallback): boolean;
   setDefaultEncoding(encoding: string): this;
-  end(): void;
-  end(chunk: any & T, cb?: Function): void;
-  end(chunk: any & T, encoding?: any, cb?: Function): void;
+  end(): this;
+  end(chunk: any & T, cb?: Function): this;
+  end(chunk: any & T, encoding?: any, cb?: Function): this;
 }
 
 export interface ObjectWritable<T> extends IntermediateObjectWritable<T> {
@@ -46,20 +46,7 @@ export interface ObjectWritable<T> extends IntermediateObjectWritable<T> {
   write(chunk: T, cb?: Function): boolean;
   write(chunk: T, encoding?: any, cb?: Function): boolean;
   setDefaultEncoding(encoding: string): this;
-  end(): void;
-  end(chunk: T, cb?: Function): void;
-  end(chunk: T, encoding?: any, cb?: Function): void;
+  end(): this;
+  end(chunk: T, cb?: Function): this;
+  end(chunk: T, encoding?: any, cb?: Function): this;
 }
-
-export type ObjectDuplex<T, U> = {
-  read(size?: number): U;
-
-  _write(chunk: T, encoding: string, callback: Function): void;
-  write(chunk: T, cb?: Function): boolean;
-  write(chunk: T, encoding?: any, cb?: Function): boolean;
-  end(): void;
-  end(chunk: T, cb?: Function): void;
-  end(chunk: T, encoding?: any, cb?: Function): void;
-} & Duplex &
-  ObjectWritable<T> &
-  ObjectReadable<U>;

--- a/packages/grpc-js/src/server-call.ts
+++ b/packages/grpc-js/src/server-call.ts
@@ -236,7 +236,7 @@ export class ServerWritableStreamImpl<RequestType, ResponseType>
       this.trailingMetadata = metadata;
     }
 
-    super.end();
+    return super.end();
   }
 }
 
@@ -282,7 +282,7 @@ export class ServerDuplexStreamImpl<RequestType, ResponseType>
       this.trailingMetadata = metadata;
     }
 
-    super.end();
+    return super.end();
   }
 }
 
@@ -292,7 +292,6 @@ ServerDuplexStreamImpl.prototype._write =
   ServerWritableStreamImpl.prototype._write;
 ServerDuplexStreamImpl.prototype._final =
   ServerWritableStreamImpl.prototype._final;
-ServerDuplexStreamImpl.prototype.end = ServerWritableStreamImpl.prototype.end;
 
 // Unary response callback signature.
 export type sendUnaryData<ResponseType> = (

